### DISCUSSION TITLE
Realtime Database Getting Data

### DISCUSSION THREAD

**Discussion by Gegbee** [GodotNuts/GodotFirebase#185] [Q&A] [answered]
Hi,

I am currently tinkering with this plugin and I am enjoying it a lot. Saving data and deleting data from the Realtime database is very easy, but I find that receiving new data is much harder. The documentation is vague on what each event does and I was also wondering if there was a way to just get all the data in the database or see if something exists within it.

**Comment by BearDooks** [comment, marked as answer]
@Yello12 sorry it took so long to get back to you. The Realtime database updates happen in the background, so as long as you have a connection to a path, there is no need to do anything else. If that path is updated, then you will get the update. 

There is a signal that should be emitted when the path is added to our updated.

Does this answer the question?

~ Chuck

**Reply by WolfgangSenff** [reply]
@BearDooks - good answer. I'm marking it as such, but elaborating a bit.

1: To get all the data in the database, all you have to do is set a listener at the root level, make it a one-shot (you don't want to get all updates all the time, that can cause a huge power drain/data usage), which is an option available when you get a reference, and do not put any filters on it. That will ensure everything comes back. You can then store it locally and make changes to it as you see fit.

2: New data should just be from the new_data_update signal, found here: https://github.com/GodotNuts/GodotFirebase/blob/548da1f1a158a1bd4a700245faf648dd319b7671/addons/godot-firebase/database/reference.gd#L9

Patch (the signal just below it) represents updates to the position you're listening to - like in-line updates, vs new is completely new, pushed data.

**Reply by yshunbin** [reply]
How do I call this new_data_update signal?

**Reply by WolfgangSenff** [reply]
> How do I call this new_data_update signal?

You don't! For this plugin, you connect to the signal and the plugin calls your function when data is available. See https://docs.godotengine.org/en/stable/getting_started/step_by_step/signals.html for more on how to use signals.

**Comment by exscruzme** [comment]
Hi! I know that this is an older question, but I wanted to follow up on this as it seems relevant to what I'm trying to do in my project. 

For our use case, we don't really want to keep an open connection to the database, but rather every so often just pull data at a certain path and update our local store if necessary. For this, it seems like I want to do what you've recommended in this answer and get all the data at one path, connect to its `new_data_update `signal as a one shot after getting all the data back and then disconnect from the database until we want to do this again. 

As I was debugging through, it looks like I am able to get all of the data back after initially creating the database reference, but the reference then loops through all of the results and emits a separate new_data_update signal for each object in the dictionary, which is not ideal if we are trying to just get everything back at once. Is there a way to ensure that this signal is emitted with all of the data at once? 

![image](https://user-images.githubusercontent.com/6991957/222937888-75533f71-ab8a-4098-b71e-9b28116b2e1b.png)
For example, if our data structure looks like the screenshot above, and we want all of the data to come back when the new_data_update signal is emitted (rather than the signal firing 3 separate times for each item), what should we pass to the `get_database_reference()` method? Currently I am passing the "root path" at which we would like to retrieve the data which in this example would be `/allData`. Any suggestions? 

And lastly, what's the proper way to disconnect from the database in this case?
